### PR TITLE
Convert contraction hierarchies to road-based pathfinding v2

### PR DESCRIPTION
**Pull request by dabreegster**
# Benefits

There are less directed roads than there are lanes, so there are some performance benefits to changing contraction hierarchies to operate on smaller graphs. For `huge_seattle`, the file size drops from 355MB to 339MB. Time to built all of the contraction hierarchies drops from 1001s to 724s. Both of these are related to the number of nodes in the graph -- 270k before, 112k after. I didn't measure the run-time impact for editing a few lanes and rebuilding the contraction hierarchies, but a smaller graph is likely to help there too.

# Validation

- I used temporary logging to verify that uber-turns found at the road-level path are preserved in `path_v2_to_v1`, using `south_seattle` to check.
- Since we now cache `Direction` on `Lane`, I reversed some lane directions and verified the internal direction gets flipped during map edits.
- I checked the number of cancelled trips. There's a slight increase, because the road-based pathfinder sometimes hits the case described in https://github.com/a-b-street/abstreet/issues/555#issue-822548690, with the first turn required being from a different lane than the starting lane. This'll be resolved when the simulation layer can handle exiting a driveway onto any lane, and it's not a big increase in the meantime.
- I made sure lakeslice doesn't gridlock after all of these changes, and I'm currently regenerating all other maps now too.

# Next steps

- Write the sliding window version of `path_v2_to_v1`. Actually the current Dijkstra impl isn't as slow as I thought, so this is less urgent. I will probably try to do this when I change the simulation layer to natively understand pathfinding v2, since that's the one place it'll really be used.
- Cut over the Dijkstra's and CH implementations for walking as well.
- Start introducing a new public pathfinding API and using it everywhere.
- Do the simulation-layer stuff to handle exiting a driveway onto any lane.

### CODE REVIEW

File: map_model/src/objects/road.rs [hunk 1]
```diff
@@ -434,17 +436,6 @@ impl Road {
         result
     }
 
-    /// lane must belong to this road. Offset 0 is the centermost lane on each side of a road, then
```
Comment (dabreegster) [inline map_model/src/objects/road.rs:437]: I just moved this up to be next to `offset`. I thought I could consolidate, but they calculate different things.

File: map_model/src/pathfind/dijkstra.rs [hunk 2]
```diff
@@ -124,12 +124,8 @@ fn calc_path_v2(
         |_| Duration::ZERO,
     )?;
 
-    let mut steps = Vec::new();
```
Comment (dabreegster) [inline map_model/src/pathfind/dijkstra.rs:127]: I just realized `steps` was a roundabout way of exactly recreating `path`

File: map_model/src/traversable.rs [hunk 3]
```diff
@@ -70,29 +70,32 @@ impl Position {
         }
     }
 
+    /// Given a position along a lane, find the equivalent position along another lane on the same
```
Comment (dabreegster) [inline map_model/src/traversable.rs:73]: No behavioral change here. While refactoring for `Direction`, I just made the variable names less confusing.

File: sim/src/mechanics/queue.rs [hunk 4]
```diff
@@ -240,10 +240,9 @@ impl Queue {
     /// If true, there's room and the car must actually start the turn (because the space is
     /// reserved).
     pub fn try_to_reserve_entry(&mut self, car: &Car, force_entry: bool) -> bool {
-        // If lane is already filled, then always return false, even if forced.
-        if self.reserved_length >= self.geom_len {
-            return false;
-        }
+        // If self.reserved_length >= self.geom_len, then the lane is already full. Normally we
```
Comment (dabreegster) [inline sim/src/mechanics/queue.rs:243]: Before, only 2 cars will cross SW-bound at a time, because they're not overfilling the reservation:
polite
I added this intersection (`53211694`) to the whitelist in the other file, and then 3 will go at a time. Watching it, the behavior is still ridiculous -- the throughput is so low that all of Madison backs up, then Lake Washington gets demolished too.

So I let all of the box-blocking in the world happen:
allow_blocking
It's not great that vehicles now totally get stuck in the intersection while the north/south stage is active, but in practice for this map, it works and is _much_ better than previously. We still have future work to do to get reasonable throughput, but still not have vehicles block the box this badly.

**Review by dabreegster — COMMENTED** [review]
The diff is easier to understand by individual commits

**Comment by dabreegster** [comment]
All tests done; no gridlock issues. And in fact, Poundbury once again runs!

File: sim/src/mechanics/intersection.rs [hunk 5]
```diff
@@ -1084,6 +1084,7 @@ fn allow_block_the_box(i: &Intersection) -> bool {
         || id == 1726088130
         || id == 53217946
         || id == 53223864
+        || id == 53211694
```
Comment (michaelkirk) [inline sim/src/mechanics/intersection.rs:1087]: Are these OSM Id's? My understanding is those aren't technically stable identifiers, but they don't change that much in practice. Right?

Comment (dabreegster) [reply sim/src/mechanics/intersection.rs:1087]: Correct on both. We also rely on OSM ID stability for map edits to last over time. Every now and then somebody splits  way near a traffic signal that we bake in, and I have to go manually redo it.

As far as I'm aware, there are no alternatives. :(  https://wiki.openstreetmap.org/wiki/Permanent_ID is just one proposal that never seemed to take off.

Comment (dabreegster) [reply sim/src/mechanics/intersection.rs:1087]: Also this function would ideally be smarter and detect all cases when it's better to disable blocking-the-box at an intersection. We've got one case working -- degenerate intersections -- but there's more we could capture by looking for cases where blocking an intersection couldn't possibly bother the other direction, like with:
Screenshot from 2021-04-06 16-09-46

File: map_model/src/pathfind/v2.rs [hunk 6]
```diff
@@ -64,6 +70,49 @@ pub fn path_v2_to_v1(req: PathRequest, road_steps: Vec<DirectedRoadID>, map: &Ma
...
 }
 
+fn find_uber_turns(
+    steps: &Vec<PathStep>,
+    map: &Map,
+    mut uber_turns_v2: Vec<UberTurnV2>,
+) -> Vec<UberTurn> {
+    // Pathfinding v1 needs to know the uber turns that the path crosses, for the simulation layer.
+    // Since we now construct the path in two stages, it's easiest to just reconstruct the uber
+    // turns after building the lane-based path.
```
Comment (michaelkirk) [inline map_model/src/pathfind/v2.rs:92]: > Since we now construct the path in two stages

Here's my understanding:

> we now construct the path in two stages

Once the pathfinding v2 stuff is complete, uber-turns will still be relevant, but not during path finding, which is the first stage of path construction, where Lanes are irrelevant, we're only concerned about Roads.

The second stage of path construction does know about Lanes. It is responsible for initial lane assignment and intersection traversal/dynamic lane changing. We will need to know about uber turns for this stage.

Is that right?

Comment (dabreegster) [reply map_model/src/pathfind/v2.rs:92]: This is correct. I won't know for sure until I go through all callers of the current pathfinding API, but I believe most of them just need to estimate path length, draw the path, or figure out if the path even exists at all. The only one I know of that needs the full detail of lanes is the simulation. So my plan is to have most places use a new pathfinding v2 API (that I've yet to write) that stays at the level of `DirectedRoadID`s. The first cut of the simulation would be to call this `path_v2_to_v1` function.

But we can do much better -- like storing the v2 path (road segments) and lazily calculating the next few lanes and turns as needed inside of `Router` and the equivalent for pedestrians. That lazy evaluation place can also discover the uber-turns as needed. I don't think preserving the uber turns from the road-based graph is necessary or useful; we can cache the lanes acting as uber-turn entrances and exits somewhere, and the lazy path evaluation can just do a lookup there. We use uber-turns at the sim level to reserve a bunch of turns in advance, run red lights, and fiddle with blocking-the-box behavior. It's also possible some of that logic won't last forever, and that we'll robustify intersection merging or something else.

To recap, uber-turns have two different purposes. At the map layer, they encode OSM turn restriction relations that're multiple road segments long -- things like no U-turn, where the middle segments are mapped as distinct OSM ways. The graph used for pathfinding has to be aware of this to avoid returning illegal paths. The second purpose is at the simulation layer -- we use uber-turns as a hint for making it past a cluster of intersections successfully.

Comment (dabreegster) [reply map_model/src/pathfind/v2.rs:92]: > Once the pathfinding v2 stuff is complete, uber-turns will still be relevant, but not during path finding

Ah sorry, this part is incorrect. The first purpose, "encode OSM turn restriction relations", is still true with pathfinding v2.

File: map_model/src/pathfind/vehicles.rs [hunk 7]
```diff
@@ -336,16 +363,21 @@ pub fn vehicle_cost_v2(
     params: &RoutingParams,
     map: &Map,
 ) -> Duration {
-    let mvmnt = mvmnt.get(map).unwrap();
+    // TODO Creating the consolidated polyline sometimes fails. It's rare, so just workaround
+    // temporarily by pretending the turn is 1m long.
+    let (mvmnt_length, mvmnt_turn_type) = mvmnt
```
Comment (michaelkirk) [inline map_model/src/pathfind/vehicles.rs:367]: Interesting! Do you have a test case for this one?

File: map_model/src/pathfind/v2.rs (hunk 6 above)
Comment (dabreegster) [reply map_model/src/pathfind/v2.rs:92]: https://www.openstreetmap.org/relation/12384357#map=19/47.62350/-122.29639 is an example of one of those OSM turn restriction relations

File: map_model/src/pathfind/vehicles.rs (hunk 7 above)
Comment (michaelkirk) [reply map_model/src/pathfind/vehicles.rs:367]: errr, I mean, do you have some example data for which we hit the failure case. I'm curious to see.

File: map_model/src/pathfind/v2.rs (hunk 6 above)
Comment (michaelkirk) [reply map_model/src/pathfind/v2.rs:92]: Got it! That all makes sense. Thanks for the explanation.

File: sim/src/mechanics/intersection.rs (hunk 5 above)
Comment (michaelkirk) [reply sim/src/mechanics/intersection.rs:1087]: Unless those are real cross-walks - wouldn't want to encourage blocking them. =)

**Review by michaelkirk — APPROVED** [review]
LGTM 

nice cleanup with the `lane.dir` caching!

File: sim/src/mechanics/intersection.rs (hunk 5 above)
Comment (dabreegster) [reply sim/src/mechanics/intersection.rs:1087]: Most of them aren't, but some are. #485 is still somewhere on the TODO, and eventually using the OSM data when it exists

File: map_model/src/pathfind/vehicles.rs (hunk 7 above)
Comment (dabreegster) [reply map_model/src/pathfind/vehicles.rs:367]: Screenshot from 2021-04-06 16-38-31
`Need at least two points for a PolyLine. at TurnID(Lane #50218, Lane #62664, Intersection #10306)`
This is in `south_seattle`. The error is thrown by https://github.com/a-b-street/abstreet/blob/8cfb377dfaf8ff5fdca5259130fb0c162ff190af/map_model/src/objects/turn.rs#L368. The geometry here looks super squished to begin with, so somehow the turn polylines are getting very broken

**Comment by dabreegster** [comment]
Thanks for the review! Uploading all the data, then will merge.
